File: src/main/java/com/kingmang/lazurite/runtime/UserDefinedFunction.java
package com.kingmang.lazurite.runtime;

import com.kingmang.lazurite.core.Function;
import com.kingmang.lazurite.exceptions.LzrException;
import com.kingmang.lazurite.parser.AST.Argument;
import com.kingmang.lazurite.parser.AST.Arguments;
import com.kingmang.lazurite.parser.AST.Statements.ReturnStatement;
import com.kingmang.lazurite.parser.AST.Statements.Statement;
import com.kingmang.lazurite.runtime.values.LzrNumber;
import com.kingmang.lazurite.runtime.values.LzrValue;
import lombok.AllArgsConstructor;

@AllArgsConstructor
public class UserDefinedFunction implements Function {

    public final Arguments arguments;
    public final Statement body;

    public int getArgsCount() {
        return arguments.size();
    }

    public String getArgsName(int index) {
        if (index < 0 || index >= getArgsCount()) return "";
        return arguments.get(index).getName();
    }

    @Override
    public LzrValue execute(LzrValue... values) {
        final int size = values.length;
        final int requiredArgsCount = arguments.getRequiredArgumentsCount();
        if (size < requiredArgsCount) {
            throw new LzrException("ArgumentsMismatchException ",String.format(
                    "Arguments count mismatch. Required %d, got %d", requiredArgsCount, size));
        }
        final int totalArgsCount = getArgsCount();
        if (size > totalArgsCount) {
            throw new LzrException("ArgumentsMismatchException ",String.format(
                    "Arguments count mismatch. Total %d, got %d", totalArgsCount, size));
        }

        try {
            Variables.push();
            for (int i = 0; i < size; i++) {
                Variables.define(getArgsName(i), values[i]);
            }
            // Optional args if exists
            for (int i = size; i < totalArgsCount; i++) {
                final Argument arg = arguments.get(i);
                Variables.define(arg.getName(), arg.getValueExpr().eval());
            }
            body.execute();
            return LzrNumber.ZERO;
        } catch (ReturnStatement rt) {
            return rt.getResult();
        } finally {
            Variables.pop();
        }
    }

    @Override
    public String toString() {
        if (body instanceof ReturnStatement) {
            return String.format("func%s = %s", arguments, ((ReturnStatement)body).expression);
        }
        return String.format("func%s %s", arguments, body);
    }
}

File: src/main/java/com/kingmang/lazurite/runtime/UserDefinedFunction.kt
package com.kingmang.lazurite.runtime

import com.kingmang.lazurite.core.Function
import com.kingmang.lazurite.core.checkRange
import com.kingmang.lazurite.parser.ast.Arguments
import com.kingmang.lazurite.parser.ast.statements.ReturnStatement
import com.kingmang.lazurite.parser.ast.statements.Statement
import com.kingmang.lazurite.runtime.values.LzrNumber
import com.kingmang.lazurite.runtime.values.LzrValue

open class UserDefinedFunction(
    val arguments: Arguments,
    val body: Statement
) : Function {
    fun getArgsCount(): Int =
        arguments.size()

    override fun execute(vararg args: LzrValue): LzrValue {
        val requiredArgsCount = arguments.requiredCount
        val totalArgsCount = getArgsCount()
        args.checkRange(requiredArgsCount, totalArgsCount)

        return try {
            Variables.push()
            for (i in 0 until requiredArgsCount)
                Variables.define(arguments.get(i).name, args[i])
            // Optional args if exists
            for (i in requiredArgsCount until totalArgsCount) {
                val arg = arguments.get(i)
                if (arg.valueExpr == null)
                    continue
                Variables.define(arg.name, arg.valueExpr.eval())
            }
            body.execute()
            LzrNumber.ZERO
        } catch (rt: ReturnStatement) {
            rt.result!!
        } finally {
            Variables.pop()
        }
    }

    override fun toString(): String {
        if (body is ReturnStatement) {
            return "func$arguments = ${body.expression}"
        }
        return "func$arguments $body"
    }
}
